### DISCUSSION TITLE
Week 4: July 8-12

### DISCUSSION THREAD

**Discussion by eeholmes** [SAFS-Varanasi-Internship/Summer-2024#17] [General]
Main task this week: Come up with project outline.

**Comment by eeholmes** [comment]
For Yifei:

Try India versus Africa:

India: lon 75 to 127, lat 110 to 30
Africa: Africa lon 0 to 75, lat 120 to 30

Try a larger filter, like 5 x 5

**Reply by eeholmes** [reply]
Feel free to think about other types of models if you want: https://geosmart-2023.hackweek.io/intro.html

Or compare CNN to these.

**Reply by eeholmes** [reply]
@yifeihang Ok, it is definitely possible to use CNN to improve chlorophyll estimation. Here is someone who did that:
https://zenodo.org/records/10011908

**Reply by eeholmes** [reply]
@yifeihang Here is the paper: https://essd.copernicus.org/articles/15/5281/2023/. Replicate their work and extending it would be a great project.

**Reply by eeholmes** [reply]
They used: "Sea surface temperature (SST), salinity (SAL), photosynthetically active radiation (PAR), and sea surface pressure (SSP) from reanalysis data and satellite observations were selected as the input of OCNET to correlate with the environment and phytoplankton biomass." But there is good reason to think that wind will be a better predictor than salinity. It would be interesting to create different CNN models (using their approach) with different variables and compare performance.

**Reply by eeholmes** [reply]
Other ideas:

* Compare the CNN approach of Hung et al (paper above) to the data from a spatial smoothing product: (Gap-filled DINEOF Chl-a).
* Create a tutorial or streamlit app to generate CNN predictions using the OCNET model with different variables.
*

**Reply by yifeihang** [reply]
Hi Eli, thank you for sharing this paper. I just skimmed through their methods and think it is certainly replicatable. One challenge might be the size of this CNN model, because it has more layers and way more filters comparing to my current model, but I can definitely implement it first and simplify it if needed.

**Reply by eeholmes** [reply]
Here is another paper using a similar model I think https://www.mdpi.com/2077-1312/11/4/810

**Reply by eeholmes** [reply]
There share their code: https://github.com/PeanutZom/ChloRec

**Reply by eeholmes** [reply]
* Another along these lines https://www.frontiersin.org/journals/marine-science/articles/10.3389/fmars.2023.1077623/full
* They use pytorch and share their code https://github.com/JoanaR/multi-mode-CNN-pytorch

**Comment by ShridharS19** [comment]
07/08/2024
Progress:
-Finished data analysis(PCA, LASSO, Decision Tree, Gradient Boosted Trees, Random Forest)
![DecisionTree](https://github.com/SAFS-Varanasi-Internship/Summer-2024/assets/108433641/92ae9468-aee8-4919-9c24-cb74dd31d9a7)
![Gradient_Boosted](https://github.com/SAFS-Varanasi-Internship/Summer-2024/assets/108433641/9387857e-e082-4801-9d2c-aae6ec7191d7)
![PCA](https://github.com/SAFS-Varanasi-Internship/Summer-2024/assets/108433641/c17a581f-6417-41fe-b9ba-7d22061dfa43)
![output](https://github.com/SAFS-Varanasi-Internship/Summer-2024/assets/108433641/d4d10b37-f3bd-4f40-b247-f4cf04e6ac97)
![LASSO](https://github.com/SAFS-Varanasi-Internship/Summer-2024/assets/108433641/49381b65-a388-4813-a447-fba2d9f33c1c)
- Identified that PINN via deepXDE is a potentially good generalizable approach based on data analysis
To do:
- get PINN working and results by friday

**Comment by yifeihang** [comment]
07/08/24 (Mon.)

Progress:
Tried ways to tune the model (changed variables, loss functions, input data) but didn't get an improved model

Goal:
- Try training on the African and Indian sides separately, test filter sizes
- Replicate the model in https://essd.copernicus.org/articles/15/5281/2023/  and compare its performance to my current model
- (If got time) look at other CNN architectures / other ML models, think about questions to explore for the next few weeks

**Comment by eeholmes** [comment]
07/09/24 (Tues) Eli

* Downloading ERA5 data https://github.com/google-research/arco-era5/blob/main/README.md
* Set up JupyterHub so that nodes run all the time.

**Comment by yifeihang** [comment]
07/09/24 (Tues.)

Progress:
- Tuned the ConvLSTM model by changing filter size, number of train and val data inputs, variables. Still work best with uwind vwind sst
![image](https://github.com/SAFS-Varanasi-Internship/Summer-2024/assets/166258283/929bd867-5946-4278-80f9-40fe4df3b561)

- Replicated the structure of the models in the above paper (similar to U-Net). Used 3 yrs of data for training and 1 yr for validation. Variables are real-time: uwind, vwind, sst, air_temp, flag (if there is valid CHL value), sin(time), cos(time).
![image](https://github.com/SAFS-Varanasi-Internship/Summer-2024/assets/166258283/9146dffd-e66b-44df-9bb9-c6f01c107c2a)

Goal for tomorrow:
- Build a gap-filling model - apply random clouds to CHL data with no missing values, take it as input along with other variables, let the model fill the cloud gaps

Project ideas:
- Use different models to the gap of CHL data from the clouds 
- See if the amount of missing data, seasons, etc. affect the model performances

**Comment by eeholmes** [comment]
Wednesday: July 10

Eli:

* link to Google Doc https://docs.google.com/document/d/1LbIuGtCFFLPO3MWRDquYCZ1IhFUcE8rhn6mDIATLR4Y/edit?usp=sharing
* Zotero: https://www.zotero.org/support/google_docs
*

**Comment by yifeihang** [comment]
07/10/24 (Wed.)

Progress:
- Tuned parameters for real-time prediction
- Built a basic filling-gap model (random clouds)

Goal:
- Add in days with partial CHL data (those with clouds)
- For day losses of a year plot: use different colors for CHL with cloud/without
- Brainstorm ways to get multiple small clouds/one big cloud

**Comment by eeholmes** [comment]
7/11/24

The real chloraphyll data looks like this. We were using a 'gap-filled' product. Question is can CNN, CNN+physics, tree models, create gap-filled Chl-a?

Picture came from here:
https://data.marine.copernicus.eu/product/OCEANCOLOUR_GLO_BGC_L3_MY_009_103/description

Here is the data explorer:
https://data.marine.copernicus.eu/viewer/expert?view=viewer&crs=epsg%3A4326&t=1719878400000&z=0&center=0%2C0&zoom=8.960003535063462&layers=W3sib3BhY2l0eSI6MSwiaWQiOiJ0ZW1wMSIsImxheWVySWQiOiJPQ0VBTkNPTE9VUl9HTE9fQkdDX0wzX01ZXzAwOV8xMDMvY21lbXNfb2JzLW9jX2dsb19iZ2MtcGxhbmt0b25fbXlfbDMtbXVsdGktNGttX1AxRF8yMDIzMTEvQ0hMIiwiekluZGV4IjowLCJsb2dTY2FsZSI6ZmFsc2V9XQ%3D%3D&basemap=dark

<img width="1304" alt="image" src="https://github.com/user-attachments/assets/0fdb62ea-7723-42d5-8c00-3e7f50c46ffd">

**Reply by eeholmes** [reply]
[Uploading Bug Report- cmems_obs-oc_glo_bgc-plankton_my_l4-gapfree-multi-4km_P1D.docx…]()

**Reply by eeholmes** [reply]
Bug Report: cmems_obs-oc_glo_bgc-plankton_my_l4-gapfree-multi-4km_P1D

This involves the gap-filled chlorophyll product. Here is link to the files but same behavior can be replicated in the Ocean Viewer.
URL: https://data.marine.copernicus.eu/product/OCEANCOLOUR_GLO_BGC_L4_MY_009_104/files?subdataset=cmems_obs-oc_glo_bgc-plankton_my_l4-gapfree-multi-4km_P1D_202311

 
Description

There are missing values in the gap-filled product and there appear in the same location on specific days of the year. Let's pick August 21st (randomly). Missing values will appear in the same location on every August 21 of every year 1997 to 2024. If I pick a different day, the same behavior occurs, but the location of the missing values is different. I am working in the Arabian Sea and have only tested in this location.

Steps to reproduce

Download 2 files for August 21 from https://data.marine.copernicus.eu/product/OCEANCOLOUR_GLO_BGC_L4_MY_009_104/files?subdataset=cmems_obs-oc_glo_bgc-plankton_my_l4-gapfree-multi-4km_P1D_202311&path=OCEANCOLOUR_GLO_BGC_L4_MY_009_104%2Fcmems_obs-oc_glo_bgc-plankton_my_l4-gapfree-multi-4km_P1D_202311%2F

20000821_cmems_obs-oc_glo_bgc-plankton_my_l4-gapfree-multi-4km_P1D.nc
20010821_cmems_obs-oc_glo_bgc-plankton_my_l4-gapfree-multi-4km_P1D.nc

Run this Python code to visualize

```
Import xarray as xr
ds2 = xr.open_dataset(“20000821_cmems_obs-oc_glo_bgc-plankton_my_l4-gapfree-multi-4km_P1D.nc")
ds2['CHL'].sel(time='2000-08-21', lat = slice(25, 5), lon=slice(60,80)).plot()
```
```
Import xarray as xr
ds2 = xr.open_dataset(“20010821_cmems_obs-oc_glo_bgc-plankton_my_l4-gapfree-multi-4km_P1D.nc")
ds2['CHL'].sel(time='2001-08-21', lat = slice(25, 5), lon=slice(60,80)).plot()
```
